sympy/polys/rings.py

```diff
@@ -3005,3 +3027,243 @@ def sqf_list(f, all=False):
+            if i >= 2:
+                break
+            gcd_monom = tuple(map(min, zip(monom_gcd, monom)))
+            if gcd_monom == zero_monom:
+                return p, None
+
+        d = ring({monom_gcd: ring.domain.one})
+        _d = degree_list(d.as_expr())
+        print(_d)
+        p_result = []
+        for pi in p:
+            pi = degree_list(pi.as_expr())
+            print(pi)
+            p_result.append(ring.monomial_ldiv(pi, _d))
+        p_result = p
+        return p, d
+
+def gcd_ground(coeffs, domain):
+    """
+    Compute the greatest common divisor (GCD) of elements in the ground domain.
+
+    Examples
+    ========
+
+    >>> from sympy import ZZ
+    >>> coeffs = [3, 12, 6]
+    >>> gcd_ground(coeffs, ZZ)
+    3
+
+    """
+    coeffs = list(coeffs)
+    gcd = domain.gcd
+    d = coeffs[0]
+
+    for coeff in coeffs[1:]:
+        d = gcd(d, coeff)
+
+        if d == domain.one:
+            break
+
+    return d
+
+def gcd_coeffs(p):
+    """
+    Simplify a list of polynomials whose gcd is wanted. Returns a possibly
+    longer list of simpler polynomials having the same gcd as the input.
+
+    This is done by eliminating symbols that can not be part of the gcd
+    because they do not appear in each item of the input. In the output
+    list, all items have exactly the same symbols. The set of those symbols
+    is also returned.
+
+    Examples
+    ========
+
+    >>> from sympy import ring, ZZ
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = x**2 - y**2
+    >>> g = x - y
+    >>> p = [f, g]
+    >>> gcd_coeffs(p)
+    ([1], None)
+
+    """
+    all_coeffs = p
+
+    while True:
+        # Quick exits are most efficient if we start from the simplest polys
+        p = sorted(set(all_coeffs), key=len)
+
+        # Find the intersection of symbols for each poly:
+        common = p[0].main_variable()
+        allsame = True
+        for pi in p[1:]:
+
+            # Quick exit
+            if not common:
+                R = p[0].ring
+                K = R.domain
+                gcd = gcd_terms(p, R, K)
+                return [gcd], None
+
+            syms = pi.main_variable()
+            if allsame and syms != common:
+                allsame = False
+            common &= syms
+
+        # The loop is complete if they all have the same symbols.
+        if allsame:
+            return p, common
+
+        # Extract coefficients as polys containing only the common symbols.
+        all_coeffs = []
+        for i, pi in enumerate(p):
+            coeffs_i = pi.coeff_wrt(pi.main_variable - common)
+            all_coeffs.extend(coeffs_i)
+
+            # Quick exit:
+            if len(coeffs_i) == 1:
+                R = p[0].ring
+                K = R.domain
+                gcd = gcd_terms((all_coeffs + p[i+1:]), R, K)
+                return [gcd], None
+
+def gcd_sparse(p):
+    """
+    Returns the greatest common divisor (GCD) of a set of polynomials.
```

Comment:
Maybe it would be better to call this `ngcd_prs`. It is basically just the `gcd_prs` function but for `n` polys instead of `2` polys. Although you would want to use this instead of `gcd_prs` even just for 2 polys.

Perhaps this should be called `gcd_prs` and the other function should be called `_gcd_prs` as the base case that actually just implements the subresultant PRS algorithm.